src/System.Security.Cryptography.Encryption/tests/CryptoStream.cs

```diff
@@ -0,0 +1,216 @@
+                encryptStream.Write(toWrite, toWrite.Length / 2, toWrite.Length - (toWrite.Length / 2));
+                Assert.False(encryptStream.HasFlushedFinalBlock);
+
+                // Write one byte at a time
+                for (int i = 0; i < toWrite.Length; i++)
+                {
+                    encryptStream.WriteByte(toWrite[i]);
+                }
+                Assert.False(encryptStream.HasFlushedFinalBlock);
+
+                // Write async
+                encryptStream.WriteAsync(toWrite, 0, toWrite.Length).GetAwaiter().GetResult();
+                Assert.False(encryptStream.HasFlushedFinalBlock);
+
+                // Flush (nops)
+                encryptStream.Flush();
+                encryptStream.FlushAsync().GetAwaiter().GetResult();
+
+                encryptStream.FlushFinalBlock();
+                Assert.Throws<NotSupportedException>(() => encryptStream.FlushFinalBlock());
+                Assert.True(encryptStream.HasFlushedFinalBlock);
+
+                Assert.True(stream.Length > 0);
+            }
+
+            stream = new MemoryStream(stream.ToArray()); // CryptoStream.Dispose disposes the stream
+            using (CryptoStream decryptStream = new CryptoStream(stream, decryptor, CryptoStreamMode.Read))
+            {
+                Assert.False(decryptStream.CanWrite);
+                Assert.True(decryptStream.CanRead);
+                Assert.False(decryptStream.CanSeek);
+                Assert.False(decryptStream.HasFlushedFinalBlock);
+                Assert.Throws<NotSupportedException>(() => decryptStream.SetLength(1));
+                Assert.Throws<NotSupportedException>(() => decryptStream.Length);
+                Assert.Throws<NotSupportedException>(() => decryptStream.Position);
+                Assert.Throws<NotSupportedException>(() => decryptStream.Position = 0);
+                Assert.Throws<NotSupportedException>(() => decryptStream.Seek(0, SeekOrigin.Begin));
+                Assert.Throws<NotSupportedException>(() => decryptStream.Write(new byte[0], 0, 0));
+                Assert.Throws<NullReferenceException>(() => decryptStream.Read(null, 0, 0)); // No arg validation on buffer?
+                Assert.Throws<ArgumentOutOfRangeException>(() => decryptStream.Read(new byte[0], -1, 0));
+                Assert.Throws<ArgumentOutOfRangeException>(() => decryptStream.Read(new byte[0], 0, -1));
+                Assert.Throws<ArgumentOutOfRangeException>(() => decryptStream.Read(new byte[0], 0, -1));
+                Assert.Throws<ArgumentException>(() => decryptStream.Read(new byte[3], 1, 4));
+
+                using (StreamReader reader = new StreamReader(decryptStream))
+                {
+                    Assert.Equal(
+                        LoremText + LoremText + LoremText + LoremText,
+                        reader.ReadToEnd());
+                }
+            }
+
+            stream = new MemoryStream(stream.ToArray()); // CryptoStream.Dispose disposes the stream
```

Comment:
Can you add a comment to pre-suggest to the reader that we're doing the read again so that it can test async?  I was fixated more on figuring out if the previous one had tried doing anything destructive through the assert sequence.